### PR TITLE
Make directory iterators not throw exceptions

### PR DESCRIPTION
**Pull request by ravenblackx**
Commit Message: Make directory iterators not throw exceptions
Additional Description: With the file system cache using directory iterators in the eviction thread, exceptions can be a problem (e.g. if a file is deleted while the directory iterator iterates, a stat() call can be provoked to fail and throw an exception). This change removes those exceptions, and also makes the directory iterator behavior less ambiguous with respect to trailing slashes (one error message showed `path//file` which should be remedied by this).

Incidentally partially cleaned up THROW_IF_NOT_OK set of macros.
Risk Level: Maybe in some context someone would prefer an exception to be thrown?
Testing: Modified existing tests.
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: Implemented.

### CODE REVIEW

**Review by alyssawilk — COMMENTED** [review]
Love exception clean up!  the one thing I'm not clear on is if the callers of the iterator need to check status() after creating an iterator.  It looks like nothing using this (except  tests) changed?
/wait-any

**Comment by ravenblackx** [comment]
> Love exception clean up! the one thing I'm not clear on is if the callers of the iterator need to check status() after creating an iterator. It looks like nothing using this (except tests) changed? /wait-any

If it gets a bad status it effectively goes to an already-at-the-end iterator, so if you don't check status() you just treat it like a "no files" iterator (so it's still valid to use as a range-based iterator, it will just silently do nothing if the path isn't there or whatever else goes wrong).

I'll do something to actually log errors for the cache when I get to some of the other cleanups over there, but for now just not crashing is already an improvement over the current behavior where it crashes with a stack trace and somehow doesn't even include the thrown error message in the log. (Especially better in the case where it's crashing on `stat()` not on `opendir()`, where a silent failure and continue is the desired behavior!)

**Comment by alyssawilk** [comment]
My concern is if there's some existing file based config, and someone changes the directory to be non-readable, the config might now silently (with no files) where before it'd bounce.  happy to pull in API shephards and see if they have concerns or no, but otherwise I assume it's a functional change which might be problematic for the control plane.  Alternately, much as I dislike it, we could make throw-or-status a constructor option if you don't want to go through and look at current call sites?
/wait-any

**Comment by ravenblackx** [comment]
> My concern is if there's some existing file based config, ...

Ah, I see - last time I looked I thought file system cache was the only non-test thing using this cursed construct, but I see there is also [DiskLayer::walkDirectory()](https://github.com/envoyproxy/envoy/blob/main/source/common/runtime/runtime_impl.cc#L413). (Though a cursory search suggests that's only called from the `DiskLayer` constructor, which is in turn only called recursively from `walkDirectory` and in `runtime_impl_test.cc`, which suggests it too might be effectively test-only.)

**Comment by ravenblackx** [comment]
Ah, no, I misread the search results and DiskLayer is constructed in `LoaderImpl::createNewSnapshot`. However, it does already have its own, better thrown exceptions for "the path doesn't exist".

**Comment by ravenblackx** [comment]
> yeah I knew it had to be used by core code somewhere otherwise we wouldn't have the OrPanic "this code is used by E-M" variant. So to make sure I understand what you're saying, you think it is used in core code but only by runtime and it already does an existence check before the iterator could fail? If that's the case I'll verify tomorrow and definitely get a second (possibly third) pair of thumbs up but will pick up reviewing

That's my understanding of the core code yeah. It *could* still potentially go down the "file was deleted while iterating over the directory was in progress and therefore stat fails" exception path (which in the modified version is "pretend that file wasn't there all along"); I don't think that one would ever have been an intentional or desirable exception though. (And even less likely it's still *possible* that you could delete the path you asked to iterate over between the runtime-specific existence check and the DirectoryIteratorImpl constructor's existence check, in which case this is a change of behavior, but again, it seems like the behavior in that unlikely racey situation being undefined is quite reasonable!)

**Comment by alyssawilk** [comment]
So looking at includes I think you're right that the iterator is only used in core code by runtime, and runtime walks it for the disk layer.  I still don't get how this isn't a behavioral change - if you are iterating over a directory which isn't readable wouldn't walking the directory previously throw and now silently just be a no-op?  I'd think you could keep the same behavior checking the stauts() as you iterate in runtime. WDYT?
I wonder if given the failure mode is silent if we should ENVOY_BUG as well.  If not maybe just comment in the .h file folks have to check the status to avoid silent failure.

**Comment by ravenblackx** [comment]
> So looking at includes I think you're right that the iterator is only used in core code by runtime, and runtime walks it for the disk layer. I still don't get how this isn't a behavioral change - if you are iterating over a directory which isn't readable wouldn't walking the directory previously throw and now silently just be a no-op? I'd think you could keep the same behavior checking the stauts() as you iterate in runtime. WDYT? I wonder if given the failure mode is silent if we should ENVOY_BUG as well. If not maybe just comment in the .h file folks have to check the status to avoid silent failure.

Ah, you're right, I didn't consider "existing but unreadable".

**Comment by ravenblackx** [comment]
> So looking at includes I think you're right that the iterator is only used in core code by runtime, and runtime walks it for the disk layer. I still don't get how this isn't a behavioral change - if you are iterating over a directory which isn't readable wouldn't walking the directory previously throw and now silently just be a no-op? I'd think you could keep the same behavior checking the stauts() as you iterate in runtime. WDYT? I wonder if given the failure mode is silent if we should ENVOY_BUG as well. If not maybe just comment in the .h file folks have to check the status to avoid silent failure.

Added comments on both DirectoryIterator and Directory, and used RELEASE_ASSERT in runtime_impl, which I think will retain the prior crashing behavior without reducing coverage or requiring me to somehow force directory operations to fail which was painful enough for the directory unit tests. (Ideally the iterator would be using the os_syscalls abstraction so that it could be fake-failed, but fixing that would be like 3 more layers of work before the cache crash bug would be addressed!)

File: source/common/runtime/runtime_impl.cc [hunk 1]
```diff
@@ -425,7 +425,11 @@ void DiskLayer::walkDirectory(const std::string& path, const std::string& prefix
   }
 
   Filesystem::Directory directory(path);
-  for (const Filesystem::DirectoryEntry& entry : directory) {
+  Filesystem::DirectoryIteratorImpl it = directory.begin();
+  RELEASE_ASSERT(it.status().ok(), it.status().ToString());
```
Comment (alyssawilk) [inline source/common/runtime/runtime_impl.cc:429]: Sorry this was a fatal failure but it should probably be a non-fatal exception going forward.  THROW_IF_NOT_OK instead?
/wait

**Review by alyssawilk — APPROVED** [review]

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**